File: app/src/main/java/mm/pndaza/tipitakamyanmar/fragment/RecentFragment.java
package mm.pndaza.tipitakamyanmar.fragment;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import mm.pndaza.tipitakamyanmar.R;
import mm.pndaza.tipitakamyanmar.adapter.RecentAdapter;
import mm.pndaza.tipitakamyanmar.database.DBOpenHelper;
import mm.pndaza.tipitakamyanmar.model.Recent;
import mm.pndaza.tipitakamyanmar.utils.MDetect;
import mm.pndaza.tipitakamyanmar.utils.Rabbit;

public class RecentFragment extends Fragment {

    public interface OnRecentItemClickListener {
        void onRecentItemClick(String bookid, int pageNumber);
    }


    private Context context;
    private RecyclerView recentListView;
    TextView emptyInfoView;
    private ArrayList<Recent> recents;
    private  OnRecentItemClickListener callbackListener;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        getActivity().setTitle(MDetect.getDeviceEncodedText(getString(R.string.recent_mm)));
        setHasOptionsMenu(true);
        return inflater.inflate(R.layout.fragment_recent, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        context = view.getContext();
        //bind view
        recentListView = view.findViewById(R.id.listView_recent);
        recentListView.setLayoutManager(new LinearLayoutManager(context));
        recentListView.addItemDecoration(new DividerItemDecoration(context, DividerItemDecoration.VERTICAL));
        applyRecentList();

        emptyInfoView = view.findViewById(R.id.empty_info);
        applyEmptyInfoView(emptyInfoView);

    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_recent, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(item.getItemId() == R.id.menu_clearAll) {
            clearRecent();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try{
            callbackListener = (OnRecentItemClickListener) context;
        } catch (ClassCastException e){
            throw new ClassCastException(context.toString() + " must implemented OnRecentItemCliclListener");

        }
    }

    private void applyRecentList() {
        recents = DBOpenHelper.getInstance(getContext()).getAllRecent();
        final RecentAdapter adapter = new RecentAdapter(recents);
        recentListView.setAdapter(adapter);
        adapter.setOnClickListener(view -> {
            RecyclerView.ViewHolder viewHolder = (RecyclerView.ViewHolder) view.getTag();
            int position = viewHolder.getAdapterPosition();
            String bookid = recents.get(position).getBookid();
            int pageNumber = recents.get(position).getPageNumber();

            Log.d("pageNumber" , ""+pageNumber);

                callbackListener.onRecentItemClick(bookid, pageNumber);
        });
    }

    private void clearRecent() {

        AlertDialog.Builder alertDialog = new AlertDialog.Builder(context,R.style.AlertDialogTheme);

        String message = "လက်တလော ကြည့်ရှုထားသည်များကို ဖယ်ရှားမှာလား";
        String comfirm = "ဖယ်ရှားမယ်";
        String cancel = "မလုပ်တော့ဘူး";
        if (!MDetect.isUnicode()) {
            message = Rabbit.uni2zg(message);
            comfirm = Rabbit.uni2zg(comfirm);
            cancel = Rabbit.uni2zg(cancel);
        }

        alertDialog.setMessage(message)
                .setCancelable(true)
                .setPositiveButton(comfirm,
                        (dialog, id) -> {
                            DBOpenHelper.getInstance(context).removeAllRecent();
                            applyRecentList();
                                applyEmptyInfoView(emptyInfoView);
                        })
                .setNegativeButton(cancel, (dialog, id) -> {
                });
        alertDialog.show();
    }

    private void applyEmptyInfoView(TextView emptyInfoView){

        String info = getString(R.string.recent_empty);
        if (!MDetect.isUnicode()) {
            info = Rabbit.uni2zg(info);
        }
        emptyInfoView.setText(info);
        emptyInfoView.setVisibility(recents.isEmpty() ? View.VISIBLE : View.INVISIBLE);
    }


}

File: app/src/main/java/mm/pndaza/tipitakamyanmar/fragment/RecentFragment.kt
package mm.pndaza.tipitakamyanmar.fragment

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import mm.pndaza.tipitakamyanmar.R
import mm.pndaza.tipitakamyanmar.adapter.RecentAdapter
import mm.pndaza.tipitakamyanmar.callback.SwipeToDeleteCallback
import mm.pndaza.tipitakamyanmar.database.DBOpenHelper
import mm.pndaza.tipitakamyanmar.databinding.FragmentRecentBinding
import mm.pndaza.tipitakamyanmar.model.Recent
import mm.pndaza.tipitakamyanmar.repository.RecentRepository
import mm.pndaza.tipitakamyanmar.utils.ActivityUtils
import mm.pndaza.tipitakamyanmar.utils.MDetect
import mm.pndaza.tipitakamyanmar.utils.Rabbit

class RecentFragment : Fragment() {

    private var _binding: FragmentRecentBinding? = null
    private val binding get() = _binding!!
    private lateinit var recentRepository: RecentRepository
    private lateinit var adapter: RecentAdapter

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentRecentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recentRepository = RecentRepository(DBOpenHelper.getInstance(requireContext()))
        setupRecyclerView()
        loadRecents()
        setupMenuProvider()
    }

    override fun onResume() {
        super.onResume()
        if (_binding != null) {
            refreshRecents()
        }
    }
    private fun setupRecyclerView() {
        adapter = RecentAdapter(
            mutableListOf(),
            onItemClick = { recent ->
                ActivityUtils.startReadBookActivity(
                    requireContext(),
                    recent.bookId,
                    recent.pageNumber,
                    ""
                )
            },
        )
        with(binding.listViewRecent) {
            layoutManager = LinearLayoutManager(requireContext())
            adapter = this@RecentFragment.adapter
            addItemDecoration(
                DividerItemDecoration(
                    requireContext(),
                    DividerItemDecoration.VERTICAL
                )
            )
        }

        ItemTouchHelper(
            SwipeToDeleteCallback(requireContext()) { position ->
                adapter.getRecentAt(position)?.let { recent ->
                    deleteRecent(recent, position)
                }
            }
        ).attachToRecyclerView(binding.listViewRecent)
    }

    private fun loadRecents() {
        val recents = recentRepository.getAll()
        adapter.getData().clear()
        adapter.getData().addAll(recents)
        adapter.notifyDataSetChanged()
        toggleEmptyView(recents.isEmpty())
    }
    private fun refreshRecents() {
        Log.d("recent", "refreshing")
        loadRecents()
    }

    private fun toggleEmptyView(show: Boolean) {
        binding.emptyInfo.text = MDetect.getInstance().getDeviceEncodedText(getString(R.string.recent_empty))
        binding.emptyInfo.visibility = if (show) View.VISIBLE else View.GONE
    }

    private fun  deleteRecent(recent: Recent, position: Int){
        recentRepository.remove(recent.bookId)
        adapter.removeItem(position)
    }


    private fun setupMenuProvider() {
        val menuProvider = object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.menu_recent, menu)
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                return when (menuItem.itemId) {
                    R.id.menu_clearAll -> {
                        clearRecent()
                        true
                    }

                    else -> false
                }
            }
        }
        requireActivity().addMenuProvider(
            menuProvider,
            viewLifecycleOwner,
            Lifecycle.State.RESUMED
        )
    }


    private fun clearRecent() {
        val (message, confirm, cancel) =
            getLocalizedStrings("ကြည့်ဆဲစာရင်းအားလုံး ဖျက်မှာလား", "ဖျက်မယ်", "မဖျက်တော့ဘူး")


        AlertDialog.Builder(requireContext(), R.style.AlertDialogTheme)
            .setMessage(message)
            .setCancelable(true)
            .setPositiveButton(confirm) { _, _ ->
                recentRepository.removeAll()
                adapter.getData().clear()
                adapter.notifyDataSetChanged()
            }
            .setNegativeButton(cancel) { _, _ ->
                // do nothing
            }
            .show()
    }


    private fun getLocalizedStrings(
        message: String,
        confirm: String,
        cancel: String
    ): Triple<String, String, String> {

        return if (MDetect.getInstance().isUnicode()) {
            Triple(message, confirm, cancel)
        } else {
            Triple(
                Rabbit.uni2zg(message),
                Rabbit.uni2zg(confirm),
                Rabbit.uni2zg(cancel)
            )
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

}